SerializationException in awsxrayexporter if you alternate between compliant X-Ray Trace IDs

**Describe the bug**

If you configure your OTel SDK to generate AWS X-Ray compliant Trace IDs, but then later a child trace uses a non-compliant X-Ray Trace ID (i.e. W3C format), the Collector will crash due to a `SerializationException` error in the `awsxrayexporter`.

This behavior has a notable use case for AWS Lambda users. When their Lambda function starts, they may already be executing within the context of a parent trace that is tracing the Lambda function. However, they may want to change the parent of subsequent traces using the context they can pull from the `lambda_event`. (i.e. from HTTP headers found in the Lambda event). That context may not necessarily be an X-Ray compliant header.

I suspect that this error is the cause of https://github.com/aws-observability/aws-otel-lambda/issues/124

**Steps to reproduce**

Here is a quick OTel Python SDK program which first generates a trace with an X-Ray compliant trace ID, but then generates a child trace with a non-X-Ray compliant trace ID:

```python
from opentelemetry import trace as trace_api
from opentelemetry.trace import SpanKind, get_tracer

from opentelemetry.sdk import trace as trace_sdk
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
    OTLPSpanExporter,
)

from opentelemetry.sdk.extension.aws.trace import AwsXRayIdGenerator

from opentelemetry.propagators.aws import AwsXRayPropagator

trace_api.set_tracer_provider(
    trace_sdk.TracerProvider(
        id_generator=AwsXRayIdGenerator(),
        active_span_processor=BatchSpanProcessor(OTLPSpanExporter()),
    )
)

tracer = get_tracer(__name__)

parent_context = AwsXRayPropagator().extract(
    { "X-Amzn-Trace-Id": "Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Sampled=1" }
)

with tracer.start_as_current_span("A good root level span") as span:
    with tracer.start_as_current_span(
        "My ID with a non X-Ray Compatible ID", context=parent_context, kind=SpanKind.SERVER
    ) as child_span:
        print("hello")

```

The install requirements should be the following:

```
pip install \
    opentelemetry-sdk-extension-aws \
    opentelemetry-exporter-otlp-proto-grpc \
    opentelemetry-propagator-aws-xray
```

Then you would want to run the latest collector image with this command:

```
docker run \
    -p 4317:4317 \
    -e HOME=/ \
    -v $HOME/.aws:/.aws \
    -v $(pwd)/:/config \
    otel/opentelemetry-collector-contrib:latest --config /config/collector-config.yml
```

and the config is found below.

**What did you expect to see?**

No errors in the collector. Specifically, the `awsxrayexporter` should not raise an exception just because the Trace ID is incorrect. At minimum the `awsxrayexporter` should post an error like it typically does if the ID is not in X-Ray compatible format:

```
2021-12-15T07:44:17.622Z        debug   awsxrayexporter@v0.40.0/awsxray.go:54   TracesExporter  {"kind": "exporter", "name": "awsxray", "type": "awsxray", "name": "awsxray", "#spans": 1}
2021-12-15T07:44:17.623Z        debug   awsxrayexporter@v0.40.0/awsxray.go:65   Error translating span.   {"kind": "exporter", "name": "awsxray", "error": "invalid xray traceid: ff78ad1769f352d7429468f45dd27eed"}
```

**What did you see instead?**

I saw the `awsxrayexporter` crash the collector with this error:

```
2021-12-15T07:35:42.304Z        error   exporterhelper/queued_retry.go:165      Exporting failed. Try enabling retry_on_failure config option.    {"kind": "exporter", "name": "awsxray", "error": "Permanent error: SerializationException: \n\tstatus code: 400, request id: ef102cf3-23f0-490f-b5d4-8e6eaa0a849e"}
go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
        go.opentelemetry.io/collector@v0.36.0/exporter/exporterhelper/queued_retry.go:165
```

**What version did you use?**
`v0.41.0`

**What config did you use?**
Config:

```yaml
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  awsxray:
    region: us-west-2

service:
  telemetry:
    logs:
      level: debug
  pipelines:
    traces:
      receivers:
        - otlp
      exporters:
        - awsxray
```

**Environment**
OS: macOS
Compiler(if manually compiled): N/A

**Additional context**

cc @jefchien @sethAmazon

--- a/exporter/awsxrayexporter/awsxray.go
+++ b/exporter/awsxrayexporter/awsxray.go
@@ -52,27 +52,15 @@ func newTracesExporter(
 		func(ctx context.Context, td pdata.Traces) error {
 			var err error
 			logger.Debug("TracesExporter", typeLog, nameLog, zap.Int("#spans", td.SpanCount()))
-			documents := make([]*string, 0, td.SpanCount())
-			for i := 0; i < td.ResourceSpans().Len(); i++ {
-				rspans := td.ResourceSpans().At(i)
-				resource := rspans.Resource()
-				for j := 0; j < rspans.InstrumentationLibrarySpans().Len(); j++ {
-					spans := rspans.InstrumentationLibrarySpans().At(j).Spans()
-					for k := 0; k < spans.Len(); k++ {
-						document, localErr := translator.MakeSegmentDocumentString(spans.At(k), resource,
-							config.(*Config).IndexedAttributes, config.(*Config).IndexAllAttributes)
-						if localErr != nil {
-							logger.Debug("Error translating span.", zap.Error(localErr))
-							continue
-						}
-						documents = append(documents, &document)
-					}
-				}
-			}
+
+			documents := extractResourceSpans(config, logger, td)
+
 			for offset := 0; offset < len(documents); offset += maxSegmentsPerPut {
-				nextOffset := offset + maxSegmentsPerPut
-				if nextOffset > td.SpanCount() {
-					nextOffset = td.SpanCount()
+				var nextOffset int
+				if offset+maxSegmentsPerPut > len(documents) {
+					nextOffset = len(documents)
+				} else {
+					nextOffset = offset + maxSegmentsPerPut
 				}
 				input := xray.PutTraceSegmentsInput{TraceSegmentDocuments: documents[offset:nextOffset]}
 				logger.Debug("request: " + input.String())
@@ -95,6 +83,27 @@ func newTracesExporter(
 			return nil
 		}),
 	)
+}
+
+func extractResourceSpans(config config.Exporter, logger *zap.Logger, td pdata.Traces) []*string {
+	documents := make([]*string, 0, td.SpanCount())
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		rspans := td.ResourceSpans().At(i)
+		resource := rspans.Resource()
+		for j := 0; j < rspans.InstrumentationLibrarySpans().Len(); j++ {
+			spans := rspans.InstrumentationLibrarySpans().At(j).Spans()
+			for k := 0; k < spans.Len(); k++ {
+				document, localErr := translator.MakeSegmentDocumentString(spans.At(k), resource,
+					config.(*Config).IndexedAttributes, config.(*Config).IndexAllAttributes)
+				if localErr != nil {
+					logger.Debug("Error translating span.", zap.Error(localErr))
+					continue
+				}
+				documents = append(documents, &document)
+			}
+		}
+	}
+	return documents
 }
 
 func wrapErrorIfBadRequest(err *error) error {

--- a/exporter/awsxrayexporter/awsxray_test.go
+++ b/exporter/awsxrayexporter/awsxray_test.go
@@ -26,8 +26,10 @@ import (
 	"github.com/stretchr/testify/assert"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config"
 	"go.opentelemetry.io/collector/model/pdata"
 	conventions "go.opentelemetry.io/collector/model/semconv/v1.5.0"
+	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/awsutil"
 )
@@ -42,6 +44,34 @@ func TestTraceExport(t *testing.T) {
 	assert.Nil(t, err)
 }
 
+func TestXraySpanTraceResourceExtraction(t *testing.T) {
+	td := constructSpanData()
+	logger, _ := zap.NewProduction()
+	assert.Len(t, extractResourceSpans(generateConfig(), logger, td), 2, "2 spans have xay trace id")
+}
+
+func TestXrayAndW3CSpanTraceExport(t *testing.T) {
+	traceExporter := initializeTracesExporter()
+	ctx := context.Background()
+	td := constructXrayAndW3CSpanData()
+	err := traceExporter.ConsumeTraces(ctx, td)
+	assert.NotNil(t, err)
+	err = traceExporter.Shutdown(ctx)
+	assert.Nil(t, err)
+}
+
+func TestXrayAndW3CSpanTraceResourceExtraction(t *testing.T) {
+	td := constructXrayAndW3CSpanData()
+	logger, _ := zap.NewProduction()
+	assert.Len(t, extractResourceSpans(generateConfig(), logger, td), 2, "2 spans have xay trace id")
+}
+
+func TestW3CSpanTraceResourceExtraction(t *testing.T) {
+	td := constructW3CSpanData()
+	logger, _ := zap.NewProduction()
+	assert.Len(t, extractResourceSpans(generateConfig(), logger, td), 0, "0 spans have xray trace id")
+}
+
 func BenchmarkForTracesExporter(b *testing.B) {
 	traceExporter := initializeTracesExporter()
 	for i := 0; i < b.N; i++ {
@@ -54,20 +84,25 @@ func BenchmarkForTracesExporter(b *testing.B) {
 }
 
 func initializeTracesExporter() component.TracesExporter {
+	exporterConfig := generateConfig()
+	mconn := new(awsutil.Conn)
+	traceExporter, err := newTracesExporter(exporterConfig, componenttest.NewNopExporterCreateSettings(), mconn)
+	if err != nil {
+		panic(err)
+	}
+	return traceExporter
+}
+
+func generateConfig() config.Exporter {
 	os.Setenv("AWS_ACCESS_KEY_ID", "AKIASSWVJUY4PZXXXXXX")
 	os.Setenv("AWS_SECRET_ACCESS_KEY", "XYrudg2H87u+ADAAq19Wqx3D41a09RsTXXXXXXXX")
 	os.Setenv("AWS_DEFAULT_REGION", "us-east-1")
 	os.Setenv("AWS_REGION", "us-east-1")
 	factory := NewFactory()
-	config := factory.CreateDefaultConfig()
-	config.(*Config).Region = "us-east-1"
-	config.(*Config).LocalMode = true
-	mconn := new(awsutil.Conn)
-	traceExporter, err := newTracesExporter(config, componenttest.NewNopExporterCreateSettings(), mconn)
-	if err != nil {
-		panic(err)
-	}
-	return traceExporter
+	exporterConfig := factory.CreateDefaultConfig()
+	exporterConfig.(*Config).Region = "us-east-1"
+	exporterConfig.(*Config).LocalMode = true
+	return exporterConfig
 }
 
 func constructSpanData() pdata.Traces {
@@ -77,9 +112,39 @@ func constructSpanData() pdata.Traces {
 	rspans := traces.ResourceSpans().AppendEmpty()
 	resource.CopyTo(rspans.Resource())
 	ispans := rspans.InstrumentationLibrarySpans().AppendEmpty()
-	constructHTTPClientSpan().CopyTo(ispans.Spans().AppendEmpty())
-	constructHTTPServerSpan().CopyTo(ispans.Spans().AppendEmpty())
+	constructXrayTraceSpanData(ispans)
 	return traces
+}
+
+func constructW3CSpanData() pdata.Traces {
+	resource := constructResource()
+	traces := pdata.NewTraces()
+	rspans := traces.ResourceSpans().AppendEmpty()
+	resource.CopyTo(rspans.Resource())
+	ispans := rspans.InstrumentationLibrarySpans().AppendEmpty()
+	constructW3CFormatTraceSpanData(ispans)
+	return traces
+}
+
+func constructXrayAndW3CSpanData() pdata.Traces {
+	resource := constructResource()
+	traces := pdata.NewTraces()
+	rspans := traces.ResourceSpans().AppendEmpty()
+	resource.CopyTo(rspans.Resource())
+	ispans := rspans.InstrumentationLibrarySpans().AppendEmpty()
+	constructXrayTraceSpanData(ispans)
+	constructW3CFormatTraceSpanData(ispans)
+	return traces
+}
+
+func constructXrayTraceSpanData(ispans pdata.InstrumentationLibrarySpans) {
+	constructHTTPClientSpan(newTraceID()).CopyTo(ispans.Spans().AppendEmpty())
+	constructHTTPServerSpan(newTraceID()).CopyTo(ispans.Spans().AppendEmpty())
+}
+
+func constructW3CFormatTraceSpanData(ispans pdata.InstrumentationLibrarySpans) {
+	constructHTTPClientSpan(constructW3CTraceID()).CopyTo(ispans.Spans().AppendEmpty())
+	constructHTTPServerSpan(constructW3CTraceID()).CopyTo(ispans.Spans().AppendEmpty())
 }
 
 func constructResource() pdata.Resource {
@@ -97,7 +162,7 @@ func constructResource() pdata.Resource {
 	return resource
 }
 
-func constructHTTPClientSpan() pdata.Span {
+func constructHTTPClientSpan(traceID pdata.TraceID) pdata.Span {
 	attributes := make(map[string]interface{})
 	attributes[conventions.AttributeHTTPMethod] = "GET"
 	attributes[conventions.AttributeHTTPURL] = "https://api.example.com/users/junit"
@@ -107,7 +172,7 @@ func constructHTTPClientSpan() pdata.Span {
 	spanAttributes := constructSpanAttributes(attributes)
 
 	span := pdata.NewSpan()
-	span.SetTraceID(newTraceID())
+	span.SetTraceID(traceID)
 	span.SetSpanID(newSegmentID())
 	span.SetParentSpanID(newSegmentID())
 	span.SetName("/users/junit")
@@ -124,7 +189,7 @@ func constructHTTPClientSpan() pdata.Span {
 	return span
 }
 
-func constructHTTPServerSpan() pdata.Span {
+func constructHTTPServerSpan(traceID pdata.TraceID) pdata.Span {
 	attributes := make(map[string]interface{})
 	attributes[conventions.AttributeHTTPMethod] = "GET"
 	attributes[conventions.AttributeHTTPURL] = "https://api.example.com/users/junit"
@@ -135,7 +200,7 @@ func constructHTTPServerSpan() pdata.Span {
 	spanAttributes := constructSpanAttributes(attributes)
 
 	span := pdata.NewSpan()
-	span.SetTraceID(newTraceID())
+	span.SetTraceID(traceID)
 	span.SetSpanID(newSegmentID())
 	span.SetParentSpanID(newSegmentID())
 	span.SetName("/users/junit")
@@ -173,6 +238,14 @@ func newTraceID() pdata.TraceID {
 	_, err := rand.Read(r[4:])
 	if err != nil {
 		panic(err)
+	}
+	return pdata.NewTraceID(r)
+}
+
+func constructW3CTraceID() pdata.TraceID {
+	var r [16]byte
+	for i := range r {
+		r[i] = byte(rand.Intn(128))
 	}
 	return pdata.NewTraceID(r)
 }

